File: faithful_steward/storage/data_provider.py
"""DataProvider — unified CRUD interface over SQLite, mirrors iOS DataProvider."""

from datetime import datetime, date, timedelta
from typing import Optional
from ..models import Transaction, BudgetConfig, Category, InputMethod
from .database import Database


class DataProviderError(Exception):
    """DataProvider 操作異常"""
    pass


class DataProvider:
    """
    統一資料存取層 — 封裝 SQLite CRUD。
    對應 iOS DataProvider（CoreData 版）。
    """

    def __init__(self, database: Database):
        self._db = database

    # ── Transaction CRUD ──

    def add_transaction(
        self,
        amount: float,
        category: Category,
        note: str | None = None,
        method: InputMethod = InputMethod.TEXT,
        created_at: datetime | None = None,
    ) -> Transaction:
        """新增一筆記帳紀錄"""
        if amount <= 0:
            raise DataProviderError("金額必須大於 0")

        if created_at is None:
            created_at = datetime.now()

        txn = Transaction(
            amount=amount,
            category=category,
            note=note,
            input_method=method,
            created_at=created_at,
            updated_at=created_at,
        )

        self._db.conn.execute(
            """INSERT INTO transactions (id, amount, category_raw, note, input_method_raw, created_at, updated_at)
               VALUES (?, ?, ?, ?, ?, ?, ?)""",
            (
                str(txn.id),
                txn.amount,
                txn.category_raw,
                txn.note,
                txn.input_method.value,
                txn.created_at.isoformat(),
                txn.updated_at.isoformat(),
            ),
        )
        self._db.save()
        return txn

    def fetch_transactions(
        self,
        from_date: date,
        to_date: date,
        category: Category | None = None,
    ) -> list[Transaction]:
        """查詢指定日期範圍的交易紀錄（可選分類篩選）"""
        # 使用 < next_day 而非 <= to_date，因為 created_at 是完整 timestamp
        next_day = (to_date + timedelta(days=1)).isoformat()
        if category:
            rows = self._db.conn.execute(
                """SELECT * FROM transactions
                   WHERE created_at >= ? AND created_at < ?
                   AND category_raw = ?
                   ORDER BY created_at DESC""",
                (from_date.isoformat(), next_day, category.value),
            ).fetchall()
        else:
            rows = self._db.conn.execute(
                """SELECT * FROM transactions
                   WHERE created_at >= ? AND created_at < ?
                   ORDER BY created_at DESC""",
                (from_date.isoformat(), next_day),
            ).fetchall()

        return [Transaction.from_dict(dict(row)) for row in rows]

    def fetch_all_transactions(self) -> list[Transaction]:
        """查詢全部交易紀錄"""
        rows = self._db.conn.execute(
            "SELECT * FROM transactions ORDER BY created_at DESC"
        ).fetchall()
        return [Transaction.from_dict(dict(row)) for row in rows]

    def update_transaction(self, txn: Transaction) -> Transaction:
        """更新一筆交易"""
        txn.updated_at = datetime.now()
        self._db.conn.execute(
            """UPDATE transactions
               SET amount=?, category_raw=?, note=?, input_method_raw=?, updated_at=?
               WHERE id=?""",
            (
                txn.amount,
                txn.category_raw,
                txn.note,
                txn.input_method.value,
                txn.updated_at.isoformat(),
                str(txn.id),
            ),
        )
        self._db.save()
        return txn

    def delete_transaction(self, txn_id: str) -> bool:
        """刪除一筆交易"""
        cursor = self._db.conn.execute(
            "DELETE FROM transactions WHERE id=?", (txn_id,)
        )
        self._db.save()
        return cursor.rowcount > 0

    # ── BudgetConfig CRUD ──

    def fetch_budget_config(self, month_key: str) -> BudgetConfig | None:
        """查詢指定月份的預算設定"""
        row = self._db.conn.execute(
            "SELECT * FROM budget_configs WHERE month_key=?",
            (month_key,),
        ).fetchone()

        if row is None:
            return None
        return BudgetConfig.from_dict(dict(row))

    def fetch_latest_budget_config(self) -> BudgetConfig | None:
        """查詢最新的預算設定（按 month_key 降序）"""
        row = self._db.conn.execute(
            "SELECT * FROM budget_configs ORDER BY month_key DESC LIMIT 1"
        ).fetchone()

        if row is None:
            return None
        return BudgetConfig.from_dict(dict(row))

    def save_budget_config(
        self,
        monthly_total: float,
        ratios: dict[Category, float],
        month_key: str,
    ) -> BudgetConfig:
        """儲存或更新預算設定（month_key 唯一）"""
        if monthly_total <= 0:
            raise DataProviderError("月預算總額必須大於 0")

        # 檢查總和
        total = sum(ratios.values())
        if abs(total - 1.0) > 0.001:
            raise DataProviderError(f"比例總和必須為 100%，目前為 {total * 100:.1f}%")

        existing = self.fetch_budget_config(month_key)

        if existing:
            # 更新現有
            existing.monthly_total = monthly_total
            existing.ratios = ratios
            existing.updated_at = datetime.now()

            self._db.conn.execute(
                """UPDATE budget_configs
                   SET monthly_total=?, ratios_json=?, updated_at=?
                   WHERE month_key=?""",
                (
                    existing.monthly_total,
                    existing.ratios_json,
                    existing.updated_at.isoformat(),
                    month_key,
                ),
            )
            self._db.save()
            return existing
        else:
            # 新建
            config = BudgetConfig(
                month_key=month_key,
                monthly_total=monthly_total,
                ratios=ratios,
            )

            self._db.conn.execute(
                """INSERT INTO budget_configs (id, month_key, monthly_total, ratios_json, updated_at)
                   VALUES (?, ?, ?, ?, ?)""",
                (
                    str(config.id),
                    config.month_key,
                    config.monthly_total,
                    config.ratios_json,
                    config.updated_at.isoformat(),
                ),
            )
            self._db.save()
            return config

File: tithe_budget/Core/Storage/DataProvider.swift
import Foundation
import CoreData

/// 統一資料存取層 — 封裝 CoreData CRUD
/// 對應 architecture 圖中的 DataProvider
final class DataProvider {
    private let context: NSManagedObjectContext

    init(context: NSManagedObjectContext) {
        self.context = context
    }

    // MARK: - Transaction CRUD

    func addTransaction(amount: Double, category: Category, note: String? = nil, method: InputMethod = .text) throws -> TransactionEntity {
        let txn = TransactionEntity(context: context)
        txn.id = UUID()
        txn.amount = amount
        txn.categoryRaw = category.rawValue
        txn.note = note
        txn.inputMethodRaw = method.rawValue
        txn.createdAt = Date().standardized
        txn.updatedAt = txn.createdAt
        try context.save()
        return txn
    }

    func fetchTransactions(from startDate: Date, to endDate: Date, category: Category? = nil) -> [TransactionEntity] {
        let request = TransactionEntity.fetchRequest()
        var predicates: [NSPredicate] = [
            NSPredicate(format: "createdAt >= %@ AND createdAt < %@", startDate as NSDate, endDate as NSDate)
        ]
        if let category = category {
            predicates.append(NSPredicate(format: "categoryRaw == %@", category.rawValue))
        }
        request.predicate = NSCompoundPredicate(andPredicateWithSubpredicates: predicates)
        request.sortDescriptors = [NSSortDescriptor(key: "createdAt", ascending: false)]
        return (try? context.fetch(request)) ?? []
    }

    func deleteTransaction(_ txn: TransactionEntity) throws {
        context.delete(txn)
        try context.save()
    }

    // MARK: - BudgetConfig CRUD

    func fetchBudgetConfig(monthKey: String) -> BudgetConfigEntity? {
        let request = BudgetConfigEntity.fetchRequest()
        request.predicate = NSPredicate(format: "monthKey == %@", monthKey)
        request.fetchLimit = 1
        return try? context.fetch(request).first
    }

    func fetchLatestBudgetConfig() -> BudgetConfigEntity? {
        let request = BudgetConfigEntity.fetchRequest()
        request.sortDescriptors = [NSSortDescriptor(key: "monthKey", ascending: false)]
        request.fetchLimit = 1
        return try? context.fetch(request).first
    }

    func saveBudgetConfig(monthlyTotal: Double, ratios: [Category: Double], monthKey: String) throws -> BudgetConfigEntity {
        let config = fetchBudgetConfig(monthKey: monthKey) ?? BudgetConfigEntity(context: context)
        config.id = config.id ?? UUID()
        config.monthKey = monthKey
        config.monthlyTotal = monthlyTotal

        let data = try JSONEncoder().encode(ratios.mapKeys { $0.rawValue })
        config.ratiosJSON = String(data: data, encoding: .utf8) ?? "{}"
        config.updatedAt = Date()
        try context.save()
        return config
    }
}

extension Dictionary {
    func mapKeys<T>(_ transform: (Key) -> T) -> [T: Value] {
        var result: [T: Value] = [:]
        for (key, value) in self {
            result[transform(key)] = value
        }
        return result
    }
}

extension Date {
    /// 標準化：秒數歸零（僅記錄到分鐘精度）
    var standardized: Date {
        let cal = Calendar.current
        let components = cal.dateComponents([.year, .month, .day, .hour, .minute], from: self)
        return cal.date(from: components) ?? self
    }
}
